Add `clean` block to script block as a peer to `begin`, `process`, and `end` to allow easy resource cleanup

daxian-dbw opened the pull request:
This is an alternate implementation of the `Clean` block in PowerShell. The original PR is #9900.
This implementation doesn't rely on the `Dispose` method to run the `Clean` block, but instead makes `Clean` block more like a peer to `begin`, `process`, and `end` blocks.

The updated RFC is currently under review: https://github.com/PowerShell/PowerShell-RFC/pull/294
The doc issue for this feature: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/8090
The editor syntax PR: https://github.com/PowerShell/EditorSyntax/pull/208

To-do check list:

- [x] initial design and implementation (thanks @vexx32!)
- [x] error handling proposal
- [x] new design and implementation
    - [x] regarding `SynchronousExecutionEnumerate`
    - [x] regarding steppable pipeline
    - [x] variable analysis integration for 'Clean' block
    - [x] how to handle `Exit` in `Clean` block?
      - `Exit` from within a `Clean` block terminates the execution in the `Clean` block
      - But the `ExitException` is not propagated out of the `Clean` block
    - [x] how to handle <kbd>Ctrl+c</kbd> [mimic `finally` behavior *]
    - [x] rename the new block from `cleanup` to `clean`
    - [x] semantic check for `{ clean { } }`? No
      - decide to not add the semantic check for now, maybe this should be a script analyzer rule instead.
- [x] testing
    - [x] error handling tests
    - [x] functional tests
    - [x] breakpoint in `clean` block [manually checked]
- [x] update the original accepted RFC for this feature, with details functional behavior and error handling behavior.

\* The <kbd>Ctrl+c</kbd> behavior is still pending. The [updated RFC](https://github.com/PowerShell/PowerShell-RFC/pull/294) discusses two different behaviors in the event of pipeline stopping in details in the `Pipeline Stopping Behavior` section. The current implementation in this PR mimics the `finally` clause behavior, but it keeps the changes in `Stopper` that would be needed to support non-cancellable `clean` block execution.

-----

**[Update 01/07/2025]** The changes in `Stopper` was actually removed in the commit https://github.com/PowerShell/PowerShell/pull/15177/commits/0060e0d9696de625d65cd0577efa3749e4439ac6 based on the conversation in https://github.com/PowerShell/PowerShell/pull/15177#discussion_r696128970, because the decision was to follow the `finally` behavior in the event of pipeline stopping. The changes can be found in the abovementioned commit in case it's needed.

vexx32 committed:
:sparkles: Add cleanup block

vexx32 committed:
:recycle: Refactor & address code review

 - Move cleanup invocation into discrete method for CompiledScriptBlock
 - Tidy code paths in CompiledScriptBlock and DlrScriptCommandProcessor
 - Finish renaming work from dispose->cleanup
 - Rename critical region->critical section
 - Refactor disposal in CommandProcessorBase to avoid unnecessary paths

vexx32 committed:
:recycle: Refactor pipeline handling a bit

vexx32 committed:
:art: Linter fixes

vexx32 committed:
:recycle: Refactor dispose in CommandProcessorBase

 - Rename helper method.
 - Ensure DlrScriptCommandProcessor can't invoke Cleanup more than once

vexx32 committed:
:wrench: Adjust scope and error handling

- :bug: fix error duplication

vexx32 committed:
:recycle: Avoid double type check for cleanup

Moving half of the cleanup logic to DlrScriptCommandProcessor lets us
avoid the double type check/cast and handle script functions exclusively
in DlrScriptCommandProcessor.

Remaining branch for PSScriptCmdlet is retained in the base class's
`Dispose()` method itself (moved up the stack one method).

vexx32 committed:
:recycle: Address Rob's feedback

- Explicit type for variable declaration
- Ensure we mark cleanup as completed if there is no cleanup block.
- Check depth before taking the lock

vexx32 committed:
:recycle: Move duplicated scope handling to method

This method could also be utilised in all the Begin/Process/End paths.
However, they each have a very slightly different version of the code,
do things in different orders, and split up certain parts of the code.

We can revisit whether we want to refactor those code paths to use the
new method in future, but for now this avoids duplicating the code from
here to DlrScriptCommandProcessor, which would introduce complications.

vexx32 committed:
:recycle: Refactor critical section logic

With some review changes, it appears we're able to use Monitor methods
instead of a semaphore as was previously required.

Also added some comments.

vexx32 committed:
:recycle: Address Dongbo's review comments

daxian-dbw committed:
Refactor the design of 'Clean' block implementation

daxian-dbw committed:
Revert a few unneeded changes

daxian-dbw committed:
Fix build and avoid using 'CriticalSection' code for now

iSazonov commented:
> * Terminating error when any of 'Begin/Process/End' throws non-terminating error
> * Terminating error when any of 'Begin/Process/End' throws terminating error (duplicate error reports in $Error)

It is not clear after previous two:

> * Non-terminating error when any of 'Begin/Process/End' throws terminating error (duplicate error reports in $Error)
>   `function ooo { [cmdletbinding()]param() end { throw 'err'; 3 } cleanup { "Yeah!"; gcm blah; Write-Host 'host-4' } }`
>   `function ooo { [cmdletbinding()]param() end { gcm eee -ErrorAction Stop; 3 } cleanup { "Yeah!"; gcm blah; Write-Host 'host-4' } }`
> * Non-terminating error when any of 'Begin/Process/End' throws non-terminating error
>   `function ooo { [cmdletbinding()]param() process { gcm eee; 3 } cleanup { "Yeah!"; gcm blah; Write-Host 'host-4' } }`
>   ooo -ErrorVariable lv

iSazonov commented:
> How to handle Exit in 'Clean'?

It would be amazing for cmdlets to terminate whole script or process but acceptable for script blocks.

daxian-dbw committed:
Fix exception catching regression

daxian-dbw committed:
Fix test failures

daxian-dbw committed:
Fix pipeline regressions

JamesWTruher commented on test/powershell/Language/Scripting/PipelineBehaviour.Tests.ps1.bak:
```diff
@@ -0,0 +1,122 @@
+            $output = $powershell.EndInvoke($asyncResult)
+
+            $output | Should -Be "PROCESS"
+            $powershell.Streams.Information[0].MessageData | Should -BeExactly "CLEANUP"
```
should probably add some checks of the other streams here and a check of the `$powershell.HadErrors` property (and perhaps some explicit tests for errors, etc)

daxian-dbw replied:
Absolutely. The tests are from the original PR #9900, and I disabled it for now because there is behavior difference in my alternative design. I will make sure add adequate tests for the error handling once the proposal is accepted by the committee.

daxian-dbw replied:
Lots of tests were added, especially error handling tests.

JamesWTruher commented on test/powershell/Language/Scripting/PipelineBehaviour.Tests.ps1.bak:
```diff
@@ -0,0 +1,122 @@
+            $powershell.Streams.Information[0].MessageData | Should -BeExactly "CLEANUP"
+        }
+    }
+}
```
Additionally, my assumption is that `CLEAN` is run even in this case:
```powershell
BEGIN {
    "BEGIN"
}
PROCESS {
    "PROCESS"
    exit 1
}
END {
    "END"
}
CLEAN {
Write-Information "CLEAN"
}
```
and that `$?` would be false

daxian-dbw replied:
This does work with the prototype. Thanks for bring this up, and I will add a test for this.
But I haven't thought too much yet about what's the right way to handle `exit` in `Clean` block.
```
[C:\]
PS:2> & { process { exit 1 } cleanup { 'clean' > C:\arena\tmp\out.txt } }

C:\>echo %errorlevel%
1

C:\>type c:\arena\tmp\out.txt
clean

C:\>
```

daxian-dbw committed:
Update the comments to make it more clear

PaulHigin reviewed (changes requested):
Still reviewing, but submitting initial comments/questions.

vexx32 commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -637,40 +637,91 @@ internal void DoComplete()
+        {
+            try
+            {
+                using (commandRuntime.AllowThisCommandToWrite(permittedToWriteToPipeline: true))
```
I could be wrong / misunderstanding but from memory this is required if you want to allow the use of verbose, debug, etc. streams from within the block.

Preventing it writing to pipeline output is done elsewhere.

daxian-dbw replied:
Preventing it writing to pipeline output is done by replacing the output pipe with a null pipe when before running the `Begin`, `Process` and `End` clauses.

daxian-dbw commented on src/System.Management.Automation/engine/CommandInfo.cs:
```diff
@@ -529,7 +529,7 @@ private void GetMergedCommandParameterMetadata(out MergedCommandParameterMetadat
                 processor = scriptCommand != null
                     ? new CommandProcessor(scriptCommand, _context, useLocalScope: true, fromScriptFile: false,
                         sessionState: scriptCommand.ScriptBlock.SessionStateInternal ?? Context.EngineSessionState)
-                    : new CommandProcessor((CmdletInfo)this, _context) { UseLocalScope = true };
```
This is because `UseLocalScope` is only used for script commands (namely only when `scriptCommand != null`). It's not used at all when `CommandProcessor` is constructed with `CmdletInfo`. I spent some time trying to understand why `{ UseLocalScope = true }` is used here, and think it's probably better to remove it since `UseLocalScope` is meaningless for `CmdletInfo`, so that we don't spend more time in future trying to understand this code again.

iSazonov replied:
Maybe update the comment in UseLocalScope definition?
```
        /// <summary>
        /// For commands that use the scope stack, if this flag is
        /// true, don't create a new scope when running this command.
        /// </summary>
```

daxian-dbw commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -87,6 +91,11 @@ internal bool AddedToPipelineAlready
+        /// <summary>
+        /// Gets whether the command has a 'Clean' block defined.
+        /// </summary>
+        internal bool HasCleanBlock { get; }
```
I think the `Cleanup` block should be renamed to `Clean` block. `Begin/Process/End` are all verbs, and mapping to `BeginProcessing/ProcessRecord/EndProcessing`. So I think this new block should be mapped to `CleanResource`, and have `Clean` as its name. Of course, this will be a decision from the committee. I will make sure the name is consistent once it's decided what we will use.

daxian-dbw commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -371,13 +380,10 @@ internal void RestorePreviousScope()
 
             Context.EngineSessionState = _previousCommandSessionState;
 
-            if (_previousScope != null)
```
The methods `SetCurrentScopeToExecutionScope` and `RestorePreviousScope` are always used in pairs -- `SetCurrentScopeToExecutionScope` being called in `try` and `RestorePreviousScope` being called in `finally`.

`_previousScope` is set in `SetCurrentScopeToExecutionScope` and restored in `RestorePreviousScope`. So `_previousScope` cannot be `null` unless it's set so in `SetCurrentScopeToExecutionScope` (`_previousScope = CommandSessionState.CurrentScope;`), which is impossible (or PS is in a corrupted state if so :)).

Having `if (_previousScope != null)` here seems indicating `_previousScope` could be set to null somewhere, or the `RestorePreviousScope` could be called without pairing with `SetCurrentScopeToExecutionScope`. I think it's confusing and better to be removed.

daxian-dbw commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -452,13 +458,16 @@ internal void DoPrepare(IDictionary psDefaultParameterValues)
                 {
-                    // If we had an exception during Prepare, we're done trying to execute the command
-                    // so the scope we created needs to release any resources it hold.s
-                    CommandSessionState.RemoveScope(CommandScope);
```
This is moved to `CommandProcessorBase.Dispose(bool disposing)`, so we don't need to handle it in 3 different places (when parameter binding failed, in `End` block for commands without `Clean`, and in `Clean`).

daxian-dbw commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -637,40 +637,91 @@ internal void DoComplete()
+        internal void DoCleanup()
+        {
+            // Do not propagate exception even if the command invocation is enclosed in a TryStatement,
+            // so that the exception
```
My bad. Will complete the comment here.

daxian-dbw commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -637,40 +637,91 @@ internal void DoComplete()
+            catch (Exception e)
+            {
+                // This cmdlet threw an exception, so wrap it and bubble it up.
+                throw ManageInvocationException(e);
```
Yep, all exceptions are handled in `PipelineProcessor.Clean()`.

iSazonov replied:
Maybe add this in the code comment?

daxian-dbw commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -637,40 +637,91 @@ internal void DoComplete()
             }
         }
 
+        internal void ReportCleanupError(Exception exception)
```
It's call in [`PipelineProcessor.Clean()`](https://github.com/PowerShell/PowerShell/pull/15177/files#diff-a5b9f755ba5a72a7c39f6a710233d46fed5d734c2dad60bc3ed95be857cd2c39R728). The terminating error that was thrown when running a `Clean` block will be written to the command processor's error pipe.

daxian-dbw committed:
Complete a partial comment

daxian-dbw committed:
Changes to SteppablePipeline to support 'Clean' block

daxian-dbw committed:
Clean up the steppable pipeline changes and add more comments

JamesWTruher commented:
you suggest that errors in the clean block will be captured in `-ErrorVariable` but not generally emitted to the user as visible errors? Does that imply that the errors in the `cleanup` block will not be found in `$ERROR`? I'm a little nervous about that. I should think the clean up block runs in with implicit `SilentlyContinue` so errors are captured, but not displayed. What if the user explicitly sets `$ErrorActionPreference = "Inquire"` (or confirm), will we then prompt?

daxian-dbw commented:
> you suggest that errors in the clean block will be captured in `-ErrorVariable` but not generally emitted to the user as visible errors? Does that imply that the errors in the cleanup block will not be found in `$ERROR`? I'm a little nervous about that.

Errors in `Clean` block will be captured in `-ErrorVariable`, and also will be emitted to user as visible errors. All errors happen in `Clean` should (and can) be found in `$Error`.

> I should think the clean up block runs in with implicit `SilentlyContinue` so errors are captured, but not displayed. What if the user explicitly sets `$ErrorActionPreference = "Inquire" (or confirm)`, will we then prompt?

I personally prefer to not make the `Clean` block have a different default `ActionPreference` for error. I think it should just behave the same as other named blocks. That would also be easier for the user to know if there are any error in their `Clean` block.

@JamesWTruher The proposed error handling behavior already works with the changes in this PR, so I think it will be helpful if you can build this PR and play with it. You can also download the build artifacts from CIs, for example, the windows build can be downloaded [here](https://dev.azure.com/powershell/_apis/resources/Containers/4302657/build?itemPath=build%2Fbuild.zip). Thanks for the review and I look forward to get more feedback from you!

iSazonov commented on src/System.Management.Automation/engine/pipeline.cs:
```diff
@@ -688,10 +688,17 @@ private void DoCompleteCore(CommandProcessorBase commandRequestingUpstreamComman
+        /// <remarks>
+        /// Exception from a 'Clean' block is not allowed to propagate up and terminate the pipeline
+        /// so that other 'Clean' blocks can run without being affected. Therefore, this method will
+        /// catch and handle all exceptions inside, and it will never throw.
```
Maybe add where user can find the errors? $error, logs?

daxian-dbw replied:
The code is pretty clear that exception caught here will be written to the error stream. From a user's view, it's the same place where those errors are going to be available, $errors, error variable, redirected error streams, and etc.

iSazonov commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -452,13 +458,16 @@ internal void DoPrepare(IDictionary psDefaultParameterValues)
             }
-            catch (Exception)
+            catch (Exception e)
             {
-                if (_useLocalScope)
+                if (e is InvalidComObjectException)
```
Make sense to use `catch-when` since we re-throw others? I see below such patterns too.

daxian-dbw replied:
Yeah, actually after the change, we should replace `catch (Exception e)` with `catch (InvalidComObjectException e)` instead. Updated.

iSazonov commented on src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs:
```diff
@@ -1272,6 +1272,17 @@ public void Dispose()
+        /// <summary>
+        /// Object used to lock critical sections where script execution must occur even in stopping pipelines.
+        /// </summary>
+        private readonly object _criticalSectionLock = new object();
```
```suggestion
        private readonly object _criticalSectionLock = new();
```

iSazonov commented on src/System.Management.Automation/engine/pipeline.cs:
```diff
@@ -278,9 +277,12 @@ internal int Add(CommandProcessorBase commandProcessor)
-            if (pipelineProcessor == null) throw PSTraceSource.NewArgumentNullException(nameof(pipelineProcessor));
-            if (_redirectionPipes == null)
-                _redirectionPipes = new List<PipelineProcessor>();
+            if (pipelineProcessor is null)
+            {
+                throw PSTraceSource.NewArgumentNullException(nameof(pipelineProcessor));
+            }
```
If we want being short:
```suggestion
            pipelineProcessor ?? throw PSTraceSource.NewArgumentNullException(nameof(pipelineProcessor));
```

daxian-dbw replied:
> `pipelineProcessor ?? throw PSTraceSource.NewArgumentNullException(nameof(pipelineProcessor));`

You cannot do this, an expression cannot be used as a statement.

iSazonov commented on src/System.Management.Automation/engine/pipeline.cs:
```diff
@@ -1182,27 +1230,27 @@ private void Inject(object input, bool enumerate)
-                    if (commandProcessor == null
-                        || commandProcessor.CommandRuntime == null)
-                    {
-                        // "null command or request or ErrorOutputPipe " + i
-                        throw PSTraceSource.NewInvalidOperationException();
-                    }
+                    ValidateCommandProcessorNotNull(commandProcessor, errorMessage: null);
```
I wonder you do not change to `foreach`.

daxian-dbw replied:
Updated.

iSazonov commented on src/System.Management.Automation/engine/pipeline.cs:
```diff
@@ -1262,12 +1297,7 @@ internal void LinkPipelineErrorOutput(Pipe pipeToUse)
-                if (commandProcessor == null
-                    || commandProcessor.CommandRuntime == null)
-                {
-                    // "null command or request or ErrorOutputPipe " + i
-                    throw PSTraceSource.NewInvalidOperationException();
-                }
+                ValidateCommandProcessorNotNull(commandProcessor, errorMessage: null);
```
Th same.

daxian-dbw replied:
Updated.

SteveL-MSFT commented:
@PowerShell/powershell-committee agrees with the proposed design and suggest we move towards getting this in as an experimental feature to get user feedback

daxian-dbw committed:
Add lots of tests

daxian-dbw committed:
Add additional functional tests for steppable pipeline

daxian-dbw committed:
Rename the new block from 'cleanup' to 'clean'

daxian-dbw committed:
Fix 3 tests and a comment

daxian-dbw committed:
Update a couple test comments

daxian-dbw committed:
Add 2 more tests for 'exit' scenarios

daxian-dbw commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -589,20 +595,14 @@ internal virtual void Complete()
-                    using (ParameterBinderBase.bindingTracer.TraceScope(
-                        "CALLING EndProcessing"))
-                    {
-                        this.Command.DoEndProcessing();
-                    }
+                    this.Command.DoEndProcessing();
```
I prefer to keep the use of `using () { ... }` in this file for 2 reasons:
1. we don't actually need a variable in those cases, and creating variables just for the `using var xx = ...` syntax doesn't seem very clean. For the instances with only 1 `using () { ... }`, we can do `using var _ = ...`, but for the instances with 2 `using () { ... }`, we have to come up with 2 variable names, without really using them.
2. the nesting is not that bad.

daxian-dbw committed:
Address Ilya's comments

daxian-dbw committed:
Add some more changes:
1. not support directly invoking the 'clean' block of a script block.
2. not support a script block with the 'clean' block for '.ForEach' magic method.
3. make 'clean' block mimic the 'finally' clause
4. add a few more tests

daxian-dbw commented:
@vexx32 @iSazonov @rjmholt @PaulHigin @JamesWTruher @adityapatwardhan @SteveL-MSFT 
This PR is ready for review, the updated RFC can be found at https://github.com/PowerShell/PowerShell-RFC/pull/292.
Please take a look when you have time, and any feedback is welcome :)

iSazonov commented on test/powershell/Language/Scripting/CommandErrorHandling.Tests.ps1:
```diff
@@ -0,0 +1,470 @@
+            $failure = $null
+
+            try {
+                RunCommand -Command 'ThrowTerminatingError' -ErrorAction $ErrorAction
```
`ThrowTerminatingError`  implementation (and others test functions too) ignore `ErrorAction` parameter. Why do we need to run the test (and follows) for some test cases?

Also why do we skip Break and Inquire values.

daxian-dbw replied:
> `ThrowTerminatingError` implementation (and others test functions too) ignore `ErrorAction` parameter.

This test and the the one below it are to verify that behavior: the terminating error always stops execution without being affected by the specified error actions.

`break` breaks into debugger (even for terminating errors) and `Inquire` would also trigger interactive prompt for non-terminating errors, so both are not covered in the new tests. But I did manually test `break` and `inquire` for the `clean` block.

One more clarification: the tests in `CommandErrorHandling.Tests.ps1` are all for verifying the existing error handling behaviors in command execution. The purpose is to make sure adding the `clean` block doesn't cause a regression in the existing error handling.

iSazonov replied:
> > `ThrowTerminatingError` implementation (and others test functions too) ignore `ErrorAction` parameter.
> 
> This test and the the one below it are to verify that behavior: the terminating error always stops execution without being affected by the specified error actions.

I don't understand why we add ErrorAction parameter to RunCommand but ignore it in test function like ThrowTerminatingError. If we want really test that behavior ThrowTerminatingError should  use ErrorAction parameter too.

> One more clarification: the tests in CommandErrorHandling.Tests.ps1 are all for verifying the existing error handling behaviors in command execution. The purpose is to make sure adding the clean block doesn't cause a regression in the existing error handling.

I'd suggest to merge them in separate PR before this PR.

daxian-dbw replied:
I understand why you think that may be better, but I don't think it's necessary.

iSazonov replied:
The PR is too large and this could save a lot of time on the code review.

daxian-dbw replied:
Will move the regression protection tests to a separate PR.

iSazonov commented on test/powershell/Language/Scripting/CommandErrorHandling.Tests.ps1:
```diff
@@ -0,0 +1,470 @@
+
+    Context 'Terminating error' {
+
+        It "'ThrowTerminatingError' stops execution with 'ActionPreference.<ErrorAction>' for error action" -TestCases @(
```
Based on previous explanation I'd reword the header (and below too):
```suggestion
        It "'ThrowTerminatingError' always stops execution without being affected the explicit 'ActionPreference.<ErrorAction>' error action" -TestCases @(
```

daxian-dbw replied:
I have updated all test case names, to make them more readable.

iSazonov commented on test/powershell/Language/Scripting/CommandErrorHandling.Tests.ps1:
```diff
@@ -0,0 +1,470 @@
+            $pwsh.Streams.Verbose.Count | Should -BeExactly 1
+            $pwsh.Streams.Verbose | Should -BeExactly 'verbose-message'
+
+            ## The suppressed 'throw' exception is not written to the error stream, not sure why but it's the current behavior.
+            $pwsh.Streams.Error.Count | Should -BeExactly 0
+
+            ## The suppressed 'throw' exception is kept in '$Error'
+            $err = GetLastError
+            $err | Should -Not -BeNullOrEmpty
+            $err.FullyQualifiedErrorId | Should -BeExactly 'throw-exception'
```
What is a behavior in previous tests? Maybe add such checks there too?

As result we could combine all these tests in one with TestCases - it could be more readable to see one config table.

daxian-dbw replied:
The previous tests are dealing with terminating errors, which are surfaced as exceptions. Yeah, the exception errors should be kept in `$Error` as well, but that was not the point of those tests.

Please be noted that those tests are by no means covering every aspect of the error handling behaviors. I tried to cover most of the subtle behaviors -- some are not well known and may even be a little confusing to some extent. This is because these subtle behaviors may not be well tested, and adding those new tests may help me from introducing regression without knowing. For the more obvious/common behaviors, I hope they are already covered by existing tests.

iSazonov reviewed (commented):
I'd prefer:
1. To move regression protection tests (CommandErrorHandling.Tests.ps1) to separate PR so that we merge before the PR and ensure absence of regressions.
2. To combine the tests  in Context-s in one (per Context) with TestCases - it would be great readable.

adityapatwardhan commented:
@daxian-dbw  Please address comments from @iSazonov

daxian-dbw commented:
The regression protection tests have been moved to a separate PR #15919

iSazonov commented on src/System.Management.Automation/engine/pipeline.cs:
```diff
@@ -470,295 +466,384 @@ internal Array SynchronousExecuteEnumerate(object input)
+                ThrowFirstErrorIfExisting(logException: false);
             }
-            catch (PipelineStoppedException)
+            catch (Exception e)
             {
+                Clean();
                 DisposeCommands();
```
Wow! Did we have a bug with leaking resources?

daxian-dbw replied:
Yes, it looks like so. The `internal Array Step(object input)` method below this one always has the `catch (Exception)` block where `DisposeCommands()` is called.

vexx32 commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -1366,17 +1481,24 @@ internal override AstVisitAction InternalVisit(AstVisitor visitor)
-                action = ProcessBlock.InternalVisit(visitor);
-            if (action == AstVisitAction.Continue && EndBlock != null)
-                action = EndBlock.InternalVisit(visitor);
+            if (action == AstVisitAction.Continue)
+            {
+                _ = VisitAndShallContinue(ParamBlock) &&
+                    VisitAndShallContinue(DynamicParamBlock) &&
+                    VisitAndShallContinue(BeginBlock) &&
+                    VisitAndShallContinue(ProcessBlock) &&
+                    VisitAndShallContinue(EndBlock) &&
+                    VisitAndShallContinue(CleanBlock);
+            }
```
Heh. I remember this one. I don't think there is a good way to write this code tbh, short of shoving it entirely into a helper method with a bit more descriptive name. 😁

daxian-dbw replied:
I personally think this is concise while still readable. The discard assignment may seem a little weird at first glance, but it exists just for this purpose :)

daxian-dbw commented on src/System.Management.Automation/engine/CommandProcessorBase.cs:
```diff
@@ -637,38 +630,113 @@ internal void DoComplete()
+            //   clean
+            //   ## Note that, outer 'try/trap' doesn't affect the general exception happens in 'Clean' block.
+            //   ## so its behavior is consistent regardless of whether the command is enclosed by 'try/catch' or not.
+            //   PS:3> try { b } catch { 'outer catch' }
+            //   RuntimeException: Attempted to divide by zero.
+            //   clean
+            //
+            // Be noted that, this doesn't affect the TryStatement/TrapStatement within the 'Clean' block. Example:
+            //   ## 'try/trap' within 'Clean' block makes the general exception catch-able.
```
The `Be noted ...` line should be aligned with the `Therefore we set this property ...` line above. It's not part of the example.

daxian-dbw commented on src/System.Management.Automation/engine/ExecutionContext.cs:
```diff
@@ -846,15 +841,14 @@ internal void ResetRedirection()
-            object old = this.DollarErrorVariable;
-            ArrayList arraylist = old as ArrayList;
-            if (arraylist == null)
+            var arraylist = DollarErrorVariable as ArrayList;
+            if (arraylist is null)
```
Should use `is not`. Updated

daxian-dbw commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -1115,6 +1217,11 @@ private void SetUsingStatements(IEnumerable<UsingStatementAst> usingStatements)
         public NamedBlockAst EndBlock { get; }
 
+        /// <summary>
+        /// Gets the ast representing the dispose block for a script block, or null if no dispose block was specified.
```
Good catch. I found 3 more instances of `dispose block`, and they are all updated.

daxian-dbw commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -1778,8 +1904,7 @@ public NamedBlockAst(IScriptExtent extent, TokenKind blockName, StatementBlockAs
         {
             // Validate the block name.  If the block is unnamed, it must be an End block (for a function)
             // or Process block (for a filter).
-            if (!blockName.HasTrait(TokenFlags.ScriptBlockBlockName)
-                || (unnamed && (blockName == TokenKind.Begin || blockName == TokenKind.Dynamicparam)))
+            if (HasInvalidBlockName(blockName, unnamed))
```
The comment is accurate. blocks other than `end` and `process` have to be explicitly named.

daxian-dbw committed:
Address new comments from Ilya

daxian-dbw committed:
Update comments to make the code easier to understand

daxian-dbw committed:
Update tests to use better names

daxian-dbw commented:
@JamesWTruher @PaulHigin @rjmholt @vexx32 @adityapatwardhan @SteveL-MSFT 
Could you please review this PR when you have time? Thanks!

vexx32 reviewed (commented):
Overall looks pretty good. Beyond what I've noted here, I'm mostly just curious what folks think the desired behaviour for having the `clean` block be interruptible should be. Has the committee weighed in on that as yet?

daxian-dbw committed:
Address comments from Paul and Joel

PaulHigin reviewed (approved):
LGTM

daxian-dbw commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -856,6 +896,12 @@ public class ScriptBlockAst : Ast, IParameterMetadataProvider
                 SetParent(endBlock);
             }
 
+            if (cleanBlock != null)
```
This is to follow the existing style before and after it. I don't plan to change them all because I want to reduce the style changes in this PR.

daxian-dbw commented on src/System.Management.Automation/engine/parser/ast.cs:
```diff
@@ -1580,9 +1705,12 @@ bool IParameterMetadataProvider.UsesCmdletBinding()
         internal PipelineAst GetSimplePipeline(bool allowMultiplePipelines, out string errorId, out string errorMsg)
         {
-            if (BeginBlock != null || ProcessBlock != null || DynamicParamBlock != null)
+            if (BeginBlock != null
```
Same here. This method uses `== null` and `!= null` elsewhere, so I prefer to follow the existing style to reduce the style changes.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -891,7 +899,10 @@ internal bool SkipLogging
+            if (HasBeginBlock
+                || HasProcessBlock
+                || HasCleanBlock
+                || ast.Body.ParamBlock != null)
```
This is the only `!= null` `== null` use in this method, so I will update it.

daxian-dbw committed:
Address Aditya's comment

iSazonov commented on src/System.Management.Automation/resources/ParserStrings.resx:
```diff
@@ -1126,6 +1126,9 @@ ModuleVersion : Version of module to import. If used, ModuleName must represent
   <data name="ForEachTypeConversionFailed" xml:space="preserve">
     <value>Unable to convert input to the target type [{0}] passed to the ForEach() operator. Please check the specified type and try running your script again.</value>
   </data>
+  <data name="ForEachNotSupportCleanBlock" xml:space="preserve">
+    <value>Script block with a 'clean' block is not supported.</value>
+  </data>
```
I think we need more clear message if it is about ForEach.

daxian-dbw replied:
Good point. Message updated to bee 
```suggestion
  <data name="ForEachNotSupportCleanBlock" xml:space="preserve">
    <value>Script block with a 'clean' block is not supported by the 'ForEach' method.</value>
  </data>
```

iSazonov commented:
I set Breaking-Change label because we introduce new keyword "Clean".

daxian-dbw commented on src/System.Management.Automation/engine/CommandMetadata.cs:
```diff
@@ -1063,6 +1067,11 @@ internal string GetBeginBlock()
 
         internal string GetProcessBlock()
         {
+            // The reason we wrap scripts in 'try { } catch { throw }' (here and elsewhere) is to turn
+            // an exception that could be thrown from .NET method invocation into a terminating error
+            // that can be propagated up.
+            // By default, an exception thrown from .NET method is not terminating, but when enclosed
+            // in try/catch, it will be turned into a terminating error.
```
Good point. I added the additional data point to the issue: https://github.com/PowerShell/PowerShell/issues/6205#issuecomment-910475606

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImplicitRemotingCommands.cs:
```diff
@@ -2844,6 +2845,8 @@ private void GenerateHelperFunctions(TextWriter writer)
 
     End {{ {7} }}
 
+    Clean {{ {8} }}
+
```
I took a close look at the changes in `ImplicitRemotingCommands.cs`, and it turns out we don't need to add the `Clean` block in the generated script at all. This is because `$steppablePipeline` in the script only contains the `Invoke-Command` cmdlet. `Invoke-Command` is a binary cmdlet which doesn't support the `clean` block (`clean` block is always a non-op for binary cmdlet).

So, I will revert the changes in this file.

daxian-dbw replied:
Changed in this file reverted, except for an indentation change (it's just look so much better after changing it :)).

daxian-dbw commented on src/System.Management.Automation/engine/CommandMetadata.cs:
```diff
@@ -875,8 +875,11 @@ internal string GetProxyCommand(string helpComment, bool generateDynamicParamete
 end
 {{{5}}}
+
+clean
+{{{6}}}
 <#
-{6}
+{7}
```
Will add a test for proxy command.

daxian-dbw committed:
Revert changes in ImplicitRemotingCommands.cs because they are not needed

daxian-dbw committed:
Guard the 'clean' block behind an experimental feature

daxian-dbw commented:
@rjmholt Can you please take some time review the changes? I'd like to get your feedback before merging this PR. Thanks!

daxian-dbw commented:
Rebased the commits to resolve conflicts introduced by #16199
